### ISSUE TITLE
test_native_layer_norm_bfloat fails with "Non-divisible split with vectorization is detected. Extent: 975. Factor: 2"

### ISSUE THREAD

**Issue by casparvl** [pytorch/pytorch#76107]
### 🐛 Describe the bug

Running the following test:

```
python -m unittest test_jit_cuda_fuser.TestCudaFuser.test_native_layer_norm_bfloat -v
```

Results in the error:

```
======================================================================
ERROR: test_native_layer_norm_bfloat (test_jit_cuda_fuser.TestCudaFuser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/casparl/.local/easybuild/Centos8/2021/software/PyTorch/1.11.0-foss-2021a-CUDA-11.3.1/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 1754, in wrapper
    method(*args, **kwargs)
  File "/gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py", line 1327, in test_native_layer_norm_bfloat
    self._native_layer_norm_helper(input_shape, norm_shape, torch.bfloat16, "cuda", 1e-1)
  File "/gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py", line 1275, in _native_layer_norm_helper
    jit_o, jit_mean, jit_rstd = t_jit(x)
  File "/home/casparl/.local/easybuild/Centos8/2021/software/PyTorch/1.11.0-foss-2021a-CUDA-11.3.1/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/casparl/.local/easybuild/Centos8/2021/software/PyTorch/1.11.0-foss-2021a-CUDA-11.3.1/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 430, in prof_meth_call
    return prof_callable(meth_call, *args, **kwargs)
  File "/home/casparl/.local/easybuild/Centos8/2021/software/PyTorch/1.11.0-foss-2021a-CUDA-11.3.1/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 424, in prof_callable
    return callable(*args, **kwargs)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
RuntimeError: input_extent.value() % split_factor.value() == 0INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/executor_utils.cpp":428, please report a bug to PyTorch. Non-divisible split with vectorization is detected. Extent: 975. Factor: 2
```
I understand that vectorization won't work if the input is of size 975 and the split factor is 2, so the error 'makes sense', but I have no idea where the 975 and 2 come from. Since it is compiled code, it's hard to follow the stack trace. Note that I added some print statements, so the line numbers in the above stack trace might be 'off' by 2-5 lines compared to the unaltered version of `test_jit_cuda_fuser.py`.

The offending section of code in the test (around line 1270 or so of `pytorch/test/test_jit_cuda_fuser.py`) looks like this:

```
        x = torch.randn(shape, dtype=dtype, device=device)
        print(f'Shape of x: {x.shape}')
        t_jit = torch.jit.script(t)
        print(f'The TorchScript t_jit.graph: {t_jit.graph}')
        print(f'The TorchScript t_jit.code: {t_jit.code}')
        jit_o, jit_mean, jit_rstd = t_jit(x)
        jit_o, jit_mean, jit_rstd = t_jit(x)
        o, mean, rstd = t(x)
        self.assertEqual(o.dtype, jit_o.dtype)
```

The error occurs on the second `jit_o, jit_mean, jit_rstd = t_jit(x)`. I don't know enough about how code gets vectorized by the Torch JIT to debug this, nor do I really understand why `t_jit(x)` is called twice (and why the error only seems to occur the second time). 

Interestingly enough, the print statements show that this code section is run a few (5) times succesfully, until it finally results in the error the 6-th time. For that time, the output of the print statements is:

```
Shape of x: torch.Size([12, 15, 13, 10])
The TorchScript t_jit.graph: graph(%self : __torch__.test_jit_cuda_fuser.___torch_mangle_4.MyLayerNorm,
      %x.1 : Tensor):
  %7 : int = prim::Constant[value=10]() # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py:1264:47
  %6 : int = prim::Constant[value=13]() # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py:1264:47
  %5 : int = prim::Constant[value=15]() # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py:1264:47
  %11 : float = prim::Constant[value=1.0000000000000001e-05]() # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py:1264:88
  %o.1 : Tensor = aten::relu(%x.1) # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_cuda_fuser.py:1263:20
  %weight : Tensor = prim::GetAttr[name="weight"](%self)
  %bias : Tensor = prim::GetAttr[name="bias"](%self)
  %12 : int[] = prim::ListConstruct(%5, %6, %7)
  %13 : Tensor, %14 : Tensor, %15 : Tensor = aten::native_layer_norm(%o.1, %12, %weight, %bias, %11) # /gpfs/scratch1/shared/casparl/PyTorch/1.11.0/foss-2021a-CUDA-11.3.1/pytorch/test/test_jit_
cuda_fuser.py:1264:20
  %o.5 : (Tensor, Tensor, Tensor) = prim::TupleConstruct(%13, %14, %15)
  return (%o.5)

The TorchScript t_jit.code: def forward(self,
    x: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
  o = torch.relu(x)
  weight = self.weight
  bias = self.bias
  _0, _1, _2 = torch.native_layer_norm(o, [15, 13, 10], weight, bias, 1.0000000000000001e-05)
  return (_0, _1, _2)
```

### Versions

Note: this version of PyTorch is build from sources.

```
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 8.4 (Ootpa) (x86_64)
GCC version: (GCC) 10.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.28

Python version: 3.9.5 (default, Oct  8 2021, 23:16:43)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-305.34.2.el8_4.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.3.109
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB

Nvidia driver version: 470.57.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.3
[pip3] torch==1.11.0
[conda] Could not collect
```
Not sure why the script can't collect the cuDNN version, but it's `8.2.1.32`.

**Comment by ngimel** [comment]
cc @jjsjann123, this may be fixed on master?

**Comment by ptrblck** [comment]
This might be indeed the case as I cannot reproduce the issue on a current master build `torch==1.12.0a0+bd13bc6`:
```python
root@5146d2209d8b:/workspace/src/pytorch/test# python -m unittest test_jit_cuda_fuser.TestCudaFuser.test_native_layer_norm_bfloat -v
test_native_layer_norm_bfloat (test_jit_cuda_fuser.TestCudaFuser) ... ok

----------------------------------------------------------------------
Ran 1 test in 12.028s

OK
```

**Comment by casparvl** [comment]
Thanks for the quick responses!

Maybe a bit of context: I'm managing an HPC system, and am performing this installation for all our users. Hence, it is important for me to use an official release version. If there are real bugs, either in the code, or in the tests themselves, I'm happy to cherry-pick patches for those, but in general I'd like to stay as close as possible to the official release. If it helps, I _can_ of course try a build from master, just for the purpose of checking if the issue persists. But I'm also wondering @ptrblck: are you able to reproduce it for the 1.11.0 release? Otherwise, it might be specific to the hardware. Where you testing with similar hardware, e.g. A100's? 

In order to potentially cherry-pick a fix, I would like to better understand where this error is coming from and whether it's an issue with the test, or an _actual_ issue with the framework. As this error comes from the JIT, the backtrace is not extremely helpful. Any tips on how I can trace back where this extend of 975 and the factor of 2 come from, so that I could potentially find the underlying cause?

Considering @ptrblck couldn't reproduce on the current master, I've downloaded the latest version of `test_jit_cuda_fuser.py` from master and reran just this single test, but it still fails. So it seems at least nothing changed on the test side that would fix this.

**Comment by ptrblck** [comment]
> are you able to reproduce it for the 1.11.0 release?

Yes, I'm able to reproduce the error in `1.11.0` and get:
```
RuntimeError: input_extent.value() % split_factor.value() == 0INTERNAL ASSERT FAILED at "/workspace/src/pytorch/torch/csrc/jit/codegen/cuda/executor_utils.cpp":428, please report a bug to PyTorch. Non-divisible split with vectorization is detected. Extent: 975. Factor: 2
```
> Where you testing with similar hardware, e.g. A100's?

Yes, I'm using an A100 40GB GPU.

> So it seems at least nothing changed on the test side that would fix this.

@jjsjann123 would know which commit(s) fixed this issue. Depending on your requirement to stick to a stable release and the number of commits it might be easier to use the next `1.12.0` release, but let's wait for @jjsjann123's comment.

**Comment by jjsjann123** [comment]
We had a few PRs fixing vectorization validation in the past few months: I think the exact PR that fixes this error is: https://github.com/csarofeen/pytorch/issues/1523

**Comment by davidberard98** [comment]
@casparvl fwiw, nvfuser wasn't enabled by default in 1.11 - if you don't expect to be using nvfuser specifically, it would probably be fine to wait for 1.12

**Comment by casparvl** [comment]
Hm, seeing as how the PRs affect quite a lot of files, cherrypicking might not be a great idea. I think I'll accept the fact that this might not have been working in 1.11 and simply ignore/disable the failing tests. If nvfuser wasn't enabled by default in 1.11, I expect it would be rare for any of our users to hit any issues in this part of the framework then.

Thanks for all of your helpful replies!
